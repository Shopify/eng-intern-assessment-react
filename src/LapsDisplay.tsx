import React from 'react';
import styled from 'styled-components';

type LapsDisplayProps = {
  laps: String[];
};

const LapsDisplay: React.FC<LapsDisplayProps> = ({ laps }) => {
  return (
    <StyledLapsDisplay>
      <LapList>
        {laps.map((lap, index) => (
          <LapItem key={index}>
            <LapIndex>{`Lap #${index + 1}`}</LapIndex>
            <LapTime data-testid="lap-time">{lap}</LapTime>
          </LapItem>
        ))}
      </LapList>
    </StyledLapsDisplay>
  );
};

const StyledLapsDisplay = styled.div`
  padding: 10px;
  margin-top: 20px;
  width: 100%;
`;

const LapList = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 0;
`;

const LapItem = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-around;
  width: 100%;
  padding: 8px 16px;
  margin-bottom: 5px;
  border-bottom: 1px solid rgb(255, 255, 255, 0.4);
  &:last-child {
    border: none;
  }
`;

const LapIndex = styled.div`
  font-weight: 400;
  font-size: 1rem;
`

const LapTime = styled.div`
  font-weight: 500;
  font-size: 1rem;
`

export default LapsDisplay;
