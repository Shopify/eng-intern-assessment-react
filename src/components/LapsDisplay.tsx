import React from "react";
import functions from "../functions";

interface LapsDisplayProps {
  laps: number[];
  fastest?: number;
  slowest?: number;
}

export default function LapsDisplay({ laps, fastest, slowest }: LapsDisplayProps) {
  return (
    <div className="laps-container">
      <div className="lap-row-container">
        <div className="lap-index extra-small-text">Lap</div>
        <div className="lap-time extra-small-text">Time</div>
      </div>
      <hr className="lap-divider" />
      {laps.map((lap, index) => {
        const extraText =
          index === fastest ? "(fastest) " : index === slowest ? "(slowest) " : null;
        return (
          <div className="lap-row-container" key={index}>
            <div className="lap-index extra-small-text">{index + 1}</div>
            <div className="lap-time extra-small-text">
              {extraText && <i className="light">{extraText}</i>}&emsp;
              {functions.timeToString(lap)}
            </div>
          </div>
        );
      })}
    </div>
  );
}
