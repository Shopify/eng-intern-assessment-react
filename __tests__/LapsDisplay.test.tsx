import React from "react";
import { render } from "@testing-library/react";
import LapsDisplay from "../src/components/LapsDisplay";

describe("Test LapsDisplay", () => {
  it("renders properly", () => {
    const { container } = render(<LapsDisplay laps={[]} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders fastest and slowest", () => {
    const { queryByText } = render(<LapsDisplay laps={[1, 2, 3]} fastest={0} slowest={2} />);
    const fastestElement = queryByText("(fastest)");
    const slowestElement = queryByText("(slowest)");

    expect(fastestElement).toBeTruthy();
    expect(slowestElement).toBeTruthy();
  });

  // If only one lap, it should only display fastest, not slowest
  it("renders only fastest", () => {
    const { queryByText } = render(<LapsDisplay laps={[1]} fastest={0} slowest={0} />);
    const fastestElement = queryByText("(fastest)");
    const slowestElement = queryByText("(slowest)");

    expect(fastestElement).toBeTruthy();
    expect(slowestElement).toBeNull();
  });
});
