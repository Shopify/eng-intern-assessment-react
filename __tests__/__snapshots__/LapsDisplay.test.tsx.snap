// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LapsDisplay renders properly 1`] = `
<div
  class="laps-container"
>
  <div
    class="lap-row-container"
  >
    <div
      class="lap-index extra-small-text"
    >
      Lap
    </div>
    <div
      class="lap-time extra-small-text"
    >
      Time
    </div>
  </div>
  <hr
    class="lap-divider"
  />
</div>
`;
